app/controllers/users/sessions_controller.rb

```diff
@@ -39,6 +40,13 @@ def timeout
       redirect_to root_url
     end
 
+    protected
+
+    def require_no_authentication
```

Comment:
The way I would do this is by skipping the before_action only for the `:new` action, since that's the only one we want to override, and then add our behavior inside the `new` method:

```ruby
skip_before_action :require_no_authentication, only: [:new]

def new
  analytics.track_event(Analytics::SIGN_IN_PAGE_VISIT)

  if user_fully_authenticated?
    redirect_to after_sign_in_path_for(current_user)
    return
  else
    super
  end
end
```
I'm not sure that we need to sign the user out if they're not fully authenticated. They should be able to go back to the previous screen without having to sign back in. GitHub doesn't sign the user out, as an example.